contrib/automation_tests/test_cases/symbols_tab.py

```diff
@@ -18,21 +23,229 @@
+    of duration difference is exceeded between runs.
+
+    For example, this is used to measure symbol loading times from cold and warm caches, and verify that warm cache
+    loading is significantly faster.
+
+    Usage:
+    - Inherit your E2ETestCase from this mixin
+    - Use time.time() to record start times and duration
+    - Where needed, call `self._check_and_update_duration` to store the current duration and fail the test if applicable
+    """
+    suite = None
+
+    @abstractmethod
+    def expect_true(self, cond, msg):
+        pass
+
+    @abstractmethod
+    def expect_eq(self, left, right, msg):
+        pass
+
+    def _check_and_update_duration(self, storage_key: str, total_duration: float, expected_difference: float):
+        """
+        Compare the current duration with the previous run, and fail if it exceeds the defined bounds.
+        :param storage_key: Unique key to store the duration within the test suite, used to persist the data across
+            multiple tests (uses `self.suite.shared_data`)
+        :param total_duration: The duration of the current run
+        :param expected_difference: Expected difference in seconds. If negative, the current run must be at least
+            this much *faster* than the previous run. If positive, the current run must be at least this much *slower*.
+            Violating those conditions will result in test failure.
+        """
+        last_duration = self.suite.shared_data.get(storage_key, None)
+        self.suite.shared_data[storage_key] = total_duration
+
+        if expected_difference is not None and last_duration is not None:
+            if expected_difference < 0:
+                self.expect_true(total_duration <= last_duration + expected_difference,
+                                 "Expected symbol loading time to be at least {diff:.2f}s faster than the last run. "
+                                 "Last run duration: {last:.2f}s, current run duration: {cur:.2f}s".format(
+                                     diff=-expected_difference,
+                                     last=last_duration,
+                                     cur=total_duration))
+            else:
+                self.expect_true(total_duration >= last_duration + expected_difference,
+                                 "Expected symbol loading time to be at least {diff:.2f}s slower than the last run. "
+                                 "Last run duration: {last:.2f}s, current run duration: {cur:.2f}s".format(
+                                     diff=expected_difference,
+                                     last=last_duration,
+                                     cur=total_duration))
+
+
+ModulesLoadingResult = namedtuple("LoadingResult", ["time", "errors"])
+
+
+class ClearSymbolCache(E2ETestCase):
+    """
+    Clear the symbol cache and verify no symbol files remain.
+    """
+
```

Comment:
nit and pers pref. Remove empty line

Reply:
Our linter doesn't complain if I remove the line, but AFAIK it's standard to have the empty line after the doc comment. The reason is that this comment refers to the class, not to the following method, and this way it's clear where it belongs.